src/tdympfao.c

```diff
@@ -0,0 +1,1601 @@
+
+  mesh     = tdy->mesh;
+  cells    = mesh->cells;
+  vertices = mesh->vertices;
+
+  for (ivertex=0; ivertex<mesh->num_vertices; ivertex++) {
+    vertex = &vertices[ivertex];
+    if (vertex->num_boundary_cells == 0) {
+
+      ierr = ComputeTransmissibilityMatrixForInternalVertex(tdy, vertex, cells);
+      CHKERRQ(ierr);
+    } else {
+      if (vertex->num_internal_cells > 1) {
+        ierr = ComputeTransmissibilityMatrixForBoundaryVertex(tdy, vertex, cells);
+        CHKERRQ(ierr);
+      }
+    }
+  }
+
+  PetscFunctionReturn(0);
+
+}
+
+/* -------------------------------------------------------------------------- */
+PetscErrorCode IdentifyLocalCells(TDy tdy) {
+
+  PetscErrorCode ierr;
+  DM             dm;
+  Vec            junkVec;
+  PetscInt       junkInt;
+  PetscInt       gref;
+  PetscInt       cStart, cEnd, c;
+  TDy_cell       *cells;
+
+  PetscFunctionBegin;
+
+  dm = tdy->dm;
+  cells = tdy->mesh->cells;
+
+  // Once needs to atleast haved called a DMCreateXYZ() before using DMPlexGetPointGlobal()
+  ierr = DMCreateGlobalVector(dm, &junkVec); CHKERRQ(ierr);
+  ierr = VecDestroy(&junkVec); CHKERRQ(ierr);
+
+  ierr = DMPlexGetHeightStratum(dm, 0, &cStart, &cEnd); CHKERRQ(ierr);
+  for (c=cStart; c<cEnd; c++) {
+    ierr = DMPlexGetPointGlobal(dm,c,&gref,&junkInt); CHKERRQ(ierr);
+    if (gref>=0) {
+      cells[c].is_local = PETSC_TRUE;
+      cells[c].global_id = gref;
+    } else {
+      cells[c].is_local = PETSC_FALSE;
+      cells[c].global_id = -gref-1;
+    }
+  }
+
+  PetscFunctionReturn(0);
+
+}
+
+/* -------------------------------------------------------------------------- */
+PetscErrorCode IdentifyLocalVertices(TDy tdy) {
+
+  PetscInt       ivertex, icell, c;
+  TDy_mesh       *mesh;
+  TDy_cell       *cells;
+  TDy_vertex     *vertices;
+
+  PetscFunctionBegin;
+
+  mesh     = tdy->mesh;
+  cells    = mesh->cells;
+  vertices = mesh->vertices;
+
+  for (ivertex=0; ivertex<mesh->num_vertices; ivertex++) {
+    for (c=0; c<vertices[ivertex].num_internal_cells; c++) {
+      icell = vertices[ivertex].internal_cell_ids[c];
+      if (cells[icell].is_local) vertices[ivertex].is_local = PETSC_TRUE;
+    }
+  }
+
+  PetscFunctionReturn(0);
+
+}
+
+/* -------------------------------------------------------------------------- */
+PetscErrorCode IdentifyLocalEdges(TDy tdy) {
+
+  PetscInt iedge, icell_1, icell_2;
+  TDy_mesh *mesh;
+  TDy_cell *cells;
+  TDy_edge *edges;
+
+  PetscFunctionBegin;
+
+  mesh  = tdy->mesh;
+  cells = mesh->cells;
+  edges = mesh->edges;
+
+  for (iedge=0; iedge<mesh->num_edges; iedge++) {
+
+    if (!edges[iedge].is_internal) { // Is it a boundary edge?
+
+      // Determine the cell ID for the boundary edge
+      if (edges[iedge].cell_ids[0] != -1) icell_1 = edges[iedge].cell_ids[0];
+      else                                icell_2 = edges[iedge].cell_ids[1];
+
+      // Is the cell locally owned?
+      if (cells[icell_1].is_local) edges[iedge].is_local = PETSC_TRUE;
```

Comment:
`icell_1` is uninitialized if the above conditional is false

Reply:
The else condition should be setting `icell_1` instead of `icell_2`. It is now fixed in 7f9be5c1f4d42e8b4bafe7f50e9bd6c595a975ee